### ISSUE TITLE
Error: Unknown type: IndexedAccessType

### ISSUE THREAD

**Issue by ngaer** [lukeautry/tsoa#1122]
## Sorting

- **I'm submitting a ...**

  - [x] bug report
  - [ ] feature request
  - [x] support request

- I confirm that I
  - [x] used the [search](https://github.com/lukeautry/tsoa/search?type=Issues) to make sure that a similar issue hasn't already been submit

## Expected Behavior

The problem already was discussed in #862. Type resolver throws this error when you use a type like this:

```ts
export const MessageState: {
  pending: 'pending',
  available: 'available'
};

export type MessageState = (typeof MessageState)[keyof typeof MessageState]
```
```
Generate routes error.
 Error: Unknown type: IndexedAccessType
At: controllers/Controller.ts:16:16.
This was caused by 'type MessageState = (typeof MessageState)[keyof typeof MessageState]'
    at new GenerateMetadataError (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/exceptions.ts:6:5)
    at TypeResolver.resolve (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:318:13)
    at TypeResolver.getTypeAliasReference (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:617:120)
    at TypeResolver.getReferenceType (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:585:30)
    at TypeResolver.resolve (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:359:32)
    at /my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:107:127
    at Array.reduce (<anonymous>)
    at TypeResolver.resolve (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:106:10)
    at TypeResolver.getTypeAliasReference (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:617:120)
    at TypeResolver.getReferenceType (/my_project/node_modules/@tsoa/cli/src/metadataGeneration/typeResolver.ts:585:30)
```

## Possible Solution

Solution suggested in #862:
```ts
// file: packages/cli/src/metadataGeneration/typeResolver.ts
// ↓added
if (ts.isIndexedAccessTypeNode(this.typeNode)) {
  return new TypeResolver(this.typeNode.indexType, this.current, this.typeNode, this.context, this.referencer).resolve();
}
// ↑added

if (ts.isParenthesizedTypeNode(this.typeNode)) {
...
```
It resolves this particular case but doesn't resolve the case when using different object types with the same indexes:

```ts
const a = {
  a: 1,
  d: 4
} as const

const b = {
  a: "a",
  d: "d"
} as const

type Works = keyof typeof b // "a" | "d"
type Works = (typeof b)[keyof typeof b] // "a" | "d"
type NotWorks = (typeof a)[keyof typeof b] // 1 | 4

```

## Context (Environment)

Version of the library: 3.14.0
Version of NodeJS: 14.13.1

- Confirm you were using yarn not npm: [x]

**Comment by mvitanov** [comment]
Hey All,

I really like tsoa and am interesting in contributing to it. Would like to give this issue a try  if no one is working on it already.

Thanks!

**Comment by ngaer** [comment]
@mvitanov it would be great if you can fix it. I spent several hours on it with no luck, here's [my branch](https://github.com/ngaer/tsoa/tree/bugfix/key-of-keyword) if it can help, it has originally proposed fix and unit tests for both cases.

**Comment by relentlessricktrinidad** [comment]
Any update on this? We are currently using Prisma to generate our typescript definitions and this issue is preventing us from continuing unfortunately.

**Comment by s-chand** [comment]
Just ran into this as well using Prisma with tsoa

**Comment by WoH** [comment]
Should be fixed in #1160
